### PR TITLE
cimgui-go Migration

### PR DESCRIPTION
**Pull request by gucio321**
Hi there!
Here is my proposal of a simple way to migrate from imgui-go to cimgui-go.
I'm using regular expressions within a simple shell script.
As I finish this, I'll try to keep it up to data and uptade to the latest cimgui-go (when available)

**NOTE** - this `.sh` script is really primitive - it is just a set of `sed` calls on various files in the project.

# progress

- [x] bug in font atlas - causes strange panic
- [x] Single Window doesn't work anymore
- [x] giu theme is broken
- [x] textures does not work
   - [x] plain image
   - [x] image from file
   - [x] image from URL
- [x] some examples don't compile
- [ ] fix workflows
- [x] some `MasterWindow` methods dono't have their implementation in cimgui-go https://github.com/AllenDang/cimgui-go/issues/169, https://github.com/AllenDang/cimgui-go/issues/195
   - [x] SetSize
   - [x] SetCloseCallback
   - [x] SetDropCallback (ref:    https://github.com/AllenDang/cimgui-go/pull/145)
   - [x] SetIcon
   - [x] SetSizeLimit
   - [x] SetTitle
   - [x] SetShouldClose
   - [x] InputHandler
   - [x] GetPos
- [x] Font Atlas is broken (examples/multiplefonts) (panic when closing demo)
- [ ] broken example `examples/hugelist` (segfault)
- [ ] Implement stuff in `Plot.go` (many TODOs)
- [x] implement content scaling

<details><summary>conversion list</summary>
list is currently outdated - I'm fixing all bugs manually now

the list above shows, which files (after running migrate.sh) does not have any build errors:

- [x] Alignment.go
- [x] build_windows.go
- [x] Canvas.go
- [x] ClickableWidgets.go (**NOTE**: need to fix ImageButton)
- [x] CodeEditor.go (disabled)
- [x] Context.go
- [x] Context_test.go
- [x] CSS.go
- [x] Direction.go
- [x] doc.go
- [x] EventHandler.go
- [x] Events.go (fix Keycode first)
- [x] examples
- [x] ExtraWidgets.go
- [x] Flags.go (**NOTE** fix ColorEditFlags)
- [x] FontAtlasProsessor.go (platform and renderer dependency; cannot resolve some reference)
- [x] ImageWidgets.go (platform dependency - probably will fix manually)
- [x] InputHandler.go
- [x] InputHandler_test.go
- [x] Keycode.go
- [x] Layout.go
- [x] Layout_test.go
- [x] ListClipper.go
- [x] Markdown.go (disabled)
- [x] MasterWindow.go
- [x] MemoryEditor.go (disabled)
- [x] Msgbox.go
- [x] Plot.go
- [x] Popups.go
- [x] ProgressIndicator.go
- [x] SliderWidgets.go
- [x] SplitLayout.go
- [x] StackWidget.go
- [x] Style.go
- [x] StyleIDs.go
- [x] StyleIDs_string2enum.go
- [x] StyleIDs_string.go
- [x] StyleSetter.go
- [x] TableWidgets.go
- [x] TextureFilters.go (I'm not actually sure what these constants reffers to - no typed)
- [x] Texture.go (it's not really clear to me but I suppose giu wraps a wrong type atm - cimgui.ImGuiTextureID vs cimgui.Texture)
- [x] TextWidgets.go
- [x] Utils.go
- [x] Utils_test.go
- [x] Widgets.go (problemm with ColorEditWidget - api has changed a bit; also see TODOs in migration.sh)
- [x] Window.go (problem with platform stuff)

\* files tagged with `(disabled)` are commented-out because they are not supported by cimgui-go

</details>

### CODE REVIEW

**Comment by AllenDang** [comment]
Great progress!

**Comment by nullptr-ref** [comment]
@gucio321 Any progress on this? Would love to be able to use giu and cimgui-go in cooperation ^^

**Comment by gucio321** [comment]
@nullptr-ref yeah, will get back to it as soon as I finish cimgui-go updater, sinc it is more important imao.

**Comment by gucio321** [comment]
@AllenDang I think I'm mostly done here. The only thing I need is to rewrite MasterWindow and it _should_ be ready
could you create a new release before it is merged or give me permission to do so?

**Comment by AllenDang** [comment]
@gucio321 Thanks for your hard work! I've invited you as the collaborator for this repo.

File: FontAtlasProsessor.go [hunk 1]
```diff
@@ -300,23 +308,36 @@ func (a *FontAtlas) rebuildFontAtlas() {
...
 		}
 
 		a.extraFontMap[fontInfo.String()] = &f
 	}
 
-	fontTextureImg := fonts.TextureDataRGBA32()
-	Context.renderer.SetFontTexture(fontTextureImg)
+	// TODO
+	//fontTextureImg, _, _, _ := fonts.GetTextureDataAsRGBA32()
+	//Context.renderer.SetFontTexture(fontTextureImg)
```
Comment (gucio321) [inline FontAtlasProsessor.go:340]: @AllenDang I have a problem with thisone - what is this SetFontTexture in cimgui? Should we export something from glfw?

Comment (AllenDang) [reply FontAtlasProsessor.go:340]: Yes, you should export something from gl, the renderer.

Comment (gucio321) [reply FontAtlasProsessor.go:340]: frangly, it works without this...

**Comment by Chillance** [comment]
@gucio321 This is some impressive work. Any updates on this one? Seems very close to finish line... and it would be nice to finally leave [imgui-go](https://github.com/AllenDang/imgui-go) behind as it's not really up to date with Dear ImGUI C++ code...

**Comment by gucio321** [comment]
okey, single window works, now I need to mess up with textures :confused:

**Comment by AllenDang** [comment]
@gucio321 Maybe create a dev branch and merge it?

**Comment by gucio321** [comment]
> @gucio321 Maybe create a dev branch and merge it?

hmm I'm not sure...
I don't think someone will use such a branch. It'd be much simplier to tag current master and stop imgui-go support and fix any bugs on master

**Comment by gucio321** [comment]
@AllenDang as I mentioned in PR description we need some more methods exported in backend interface in cimgui-go before this gets merged

**Comment by gucio321** [comment]
- blocked by https://github.com/AllenDang/cimgui-go/issues/169

**Comment by AllenDang** [comment]
SetSize, SetSizeLimits, SetTItle, SetCloseCallback is ready. I need sometime to check how to SetIcon.

**Comment by gucio321** [comment]
Great! Tysm!

**Comment by gucio321** [comment]
@AllenDang I've a problem with SetCloseCallback
in [imgui](https://inkyblackness/imgui-go) `WindowCloseCallback` was `func() bool` and now we have `func(unsafe.Pointer)`
this bool was quite useful as it allowed to prevvent user from closing the window, should we drop it now?

**Comment by AllenDang** [comment]
I see, I will change it

M.Sz. ***@***.***> 于2023年7月21日周五 21:38写道：

**Comment by AllenDang** [comment]
> 

After a second look, from glfw's document, the closing process should be like this:

1. Set close callback, and the window handle will be passed to the callback.
2. Use SetShouldClose to indicate whether the close should be proceed or not.

So current implementation is sufficient.

**Comment by gucio321** [comment]
Ok i see
But it'd be better to pass backed reference instead of meaningless unsafe pointer

**Comment by AllenDang** [comment]
> Ok i see
> But it'd be better to pass backed reference instead of meaningless unsafe pointer

The unsafe pointer is the raw handle ow glfw window, I guess we could wrap it in high level package. For interact with C, I have to use the unsafe pointer to make the callback work.

**Comment by gucio321** [comment]
I've one idea, let me test it ;-)

**Comment by gucio321** [comment]
I found another bug here: `examples/multiplefonts` doesn't work at all

**Comment by gucio321** [comment]
@AllenDang I've a problem:
```go
	fontTextureImg, _, _, _ := fonts.GetTextureDataAsRGBA32()
	Context.renderer.SetFontTexture(fontTextureImg)
```
Where should I search for this SetFontTexture method? is it something from glfw?

**Comment by gucio321** [comment]
okey, I think that I'm on the right way to discover why font atlas panics. Here is what i found:
- it crashes after second RebuildFontAtlas call
- I've commented out fonts.Clear() so it doesn't clear fonts
- RebuildFontAtlas happens every frame
- adding same font second time without clearing font atlas seems to crash everything, WHY???
- Also uncommenting fonts.Clear (and replacing with ClearFonts) doesn't help
anyone any idea?

**Comment by Chillance** [comment]
I just want to give an encouragement for you all's hard work!

**Comment by AllenDang** [comment]
> @AllenDang I've a problem:
> 
> ```go
> 	fontTextureImg, _, _, _ := fonts.GetTextureDataAsRGBA32()
> 	Context.renderer.SetFontTexture(fontTextureImg)
> ```
> 
> Where should I search for this SetFontTexture method? is it something from glfw?

Let me take a look, kind of busy recently, sorry for the late reply.

**Comment by gucio321** [comment]
> GetPos is ready

Great!

> 

> > @AllenDang I've a problem:
> > ```go
> > 	fontTextureImg, _, _, _ := fonts.GetTextureDataAsRGBA32()
> > 	Context.renderer.SetFontTexture(fontTextureImg)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Where should I search for this SetFontTexture method? is it something from glfw?
> 
> Let me take a look, kind of busy recently, sorry for the late reply.

any updates?

**Comment by AllenDang** [comment]
> > GetPos is ready
> 
> Great!
> 
> > 
> 
> > > @AllenDang I've a problem:
> > > ```go
> > > 	fontTextureImg, _, _, _ := fonts.GetTextureDataAsRGBA32()
> > > 	Context.renderer.SetFontTexture(fontTextureImg)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Where should I search for this SetFontTexture method? is it something from glfw?
> > 
> > 
> > Let me take a look, kind of busy recently, sorry for the late reply.
> 
> any updates?

Give me sometime, I'm in a middle of urgent task right now.

**Comment by gucio321** [comment]
Ok no problem

**Comment by gucio321** [comment]
ok, new upstream branch gucio321/cimgui-go@dev. Waiting for:
- https://github.com/AllenDang/cimgui-go/pull/173
- https://github.com/AllenDang/cimgui-go/pull/174

**Comment by gucio321** [comment]
I found error about FontAtlas. It is:
```console
File: cimgui-go/cimgui/imgui/imgui.cpp, Line: 9887
```

**Comment by gucio321** [comment]
```go
    IM_ASSERT(SizeOfColorStack      >= g.ColorStack.Size        && "PushStyleColor/PopStyleColor Mismatch!");
```
o.k... lets say I didn't expect this?!

**Comment by gucio321** [comment]
ok, fixed
Bug in font atlas is about this:
```c++
    IM_ASSERT(g.IO.Fonts->IsBuilt()                                     && "Font Atlas not built! Make sure you called ImGui_ImplXXXX_NewFrame() function for renderer backend, which should call io.Fonts->GetTexDataAsRGBA32() / GetTexDataAsAlpha8()");
```
ref to docs:
```md
This is a current constraint of Dear ImGui (which we will lift in the future): when loading a font you need to specify which characters glyphs to load. All loaded fonts glyphs are rendered into a single texture atlas ahead of time. Calling either of io.Fonts->GetTexDataAsAlpha8(), io.Fonts->GetTexDataAsRGBA32() or io.Fonts->Build() will build the atlas. This is generally called by the Renderer backend, e.g. ImGui_ImplDX11_NewFrame() calls it. If you use custom glyphs ranges, make sure the array is persistent and available during the calls to GetTexDataAsAlpha8()/GetTexDataAsRGBA32()/Build().
```

**Comment by AllenDang** [comment]
> > GetPos is ready
> 
> Great!
> 
> > 
> 
> > > @AllenDang I've a problem:
> > > ```go
> > > 	fontTextureImg, _, _, _ := fonts.GetTextureDataAsRGBA32()
> > > 	Context.renderer.SetFontTexture(fontTextureImg)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Where should I search for this SetFontTexture method? is it something from glfw?
> > 
> > 
> > Let me take a look, kind of busy recently, sorry for the late reply.
> 
> any updates?

The related functions are already there. In glfw_backend.go, there is CreateTexture which returns a TextureID. And get the font atlas from current io, invoke it's SetTexID will do the same with SetTextureID

**Comment by gucio321** [comment]
ok, it's much better now:
image
one more problem is that I'm getting a strange panic after closng demo:
```console
free(): invalid pointer
SIGABRT: abort
PC=0x7f797b8b0844 m=4 sigcode=18446744073709551610
signal arrived during cgo execution

goroutine 5 [syscall]:
runtime.cgocall(0x6735e0, 0xc00014ff10)
	/usr/local/go/src/runtime/cgocall.go:157 +0x5c fp=0xc00014fee8 sp=0xc00014feb0 pc=0x4f495c
github.com/AllenDang/cimgui-go._Cfunc_igRunLoop(0x2911590, 0x6436e0, 0x643720, 0x643760, 0x6437e0)
	_cgo_gotypes.go:27513 +0x45 fp=0xc00014ff10 sp=0xc00014fee8 pc=0x5b9b05
github.com/AllenDang/cimgui-go.(*GLFWBackend).Run.func1(0x1?)
	/home/me/go/pkg/mod/github.com/!allen!dang/cimgui-go@v0.0.0-20230809030658-606eb8908b6c/glfw_backend.go:98 +0x9f fp=0xc00014ff70 sp=0xc00014ff10 pc=0x5be07f
github.com/AllenDang/cimgui-go.(*GLFWBackend).Run(0x0?, 0x0?)
	/home/me/go/pkg/mod/github.com/!allen!dang/cimgui-go@v0.0.0-20230809030658-606eb8908b6c/glfw_backend.go:98 +0x33 fp=0xc00014ff88 sp=0xc00014ff70 pc=0x5bdfb3
github.com/AllenDang/giu.(*MasterWindow).Run.func1()
	/home/me/git/giu/MasterWindow.go:222 +0x102 fp=0xc00014ffb0 sp=0xc00014ff88 pc=0x640c42
github.com/faiface/mainthread.Run.func1()
	/home/me/go/pkg/mod/github.com/faiface/mainthread@v0.0.0-20171120011319-8b78f0a41ae3/mainthread.go:37 +0x26 fp=0xc00014ffe0 sp=0xc00014ffb0 pc=0x63e186
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc00014ffe8 sp=0xc00014ffe0 pc=0x554941
created by github.com/faiface/mainthread.Run
	/home/me/go/pkg/mod/github.com/faiface/mainthread@v0.0.0-20171120011319-8b78f0a41ae3/mainthread.go:36 +0xb8

goroutine 1 [select, locked to thread]:
runtime.gopark(0xc00014de38?, 0x2?, 0x10?, 0x0?, 0xc00014de24?)
	/usr/local/go/src/runtime/proc.go:381 +0xd6 fp=0xc00014dcb0 sp=0xc00014dc90 pc=0x527716
runtime.selectgo(0xc00014de38, 0xc00014de20, 0x7f797b7331a8?, 0x0, 0x18?, 0x1)
	/usr/local/go/src/runtime/select.go:327 +0x7be fp=0xc00014ddf0 sp=0xc00014dcb0 pc=0x53737e
github.com/faiface/mainthread.Run(0xc000012018)
	/home/me/go/pkg/mod/github.com/faiface/mainthread@v0.0.0-20171120011319-8b78f0a41ae3/mainthread.go:42 +0x117 fp=0xc00014de68 sp=0xc00014ddf0 pc=0x63e137
github.com/AllenDang/giu.(*MasterWindow).Run(0xc000203680, 0xa1b7f0)
	/home/me/git/giu/MasterWindow.go:214 +0x72 fp=0xc00014de88 sp=0xc00014de68 pc=0x640b12
main.main()
	/home/me/git/giu/examples/multiplefonts/multiplefonts.go:46 +0x39e fp=0xc00014df80 sp=0xc00014de88 pc=0x6436be
runtime.main()
	/usr/local/go/src/runtime/proc.go:250 +0x207 fp=0xc00014dfe0 sp=0xc00014df80 pc=0x5272e7
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc00014dfe8 sp=0xc00014dfe0 pc=0x554941

goroutine 2 [force gc (idle)]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
	/usr/local/go/src/runtime/proc.go:381 +0xd6 fp=0xc000060fb0 sp=0xc000060f90 pc=0x527716
runtime.goparkunlock(...)
	/usr/local/go/src/runtime/proc.go:387
runtime.forcegchelper()
	/usr/local/go/src/runtime/proc.go:305 +0xb0 fp=0xc000060fe0 sp=0xc000060fb0 pc=0x527550
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc000060fe8 sp=0xc000060fe0 pc=0x554941
created by runtime.init.6
	/usr/local/go/src/runtime/proc.go:293 +0x25

goroutine 3 [GC sweep wait]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
	/usr/local/go/src/runtime/proc.go:381 +0xd6 fp=0xc000061780 sp=0xc000061760 pc=0x527716
runtime.goparkunlock(...)
	/usr/local/go/src/runtime/proc.go:387
runtime.bgsweep(0x0?)
	/usr/local/go/src/runtime/mgcsweep.go:278 +0x8e fp=0xc0000617c8 sp=0xc000061780 pc=0x513cce
runtime.gcenable.func1()
	/usr/local/go/src/runtime/mgc.go:178 +0x26 fp=0xc0000617e0 sp=0xc0000617c8 pc=0x509186
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc0000617e8 sp=0xc0000617e0 pc=0x554941
created by runtime.gcenable
	/usr/local/go/src/runtime/mgc.go:178 +0x6b

goroutine 4 [GC scavenge wait]:
runtime.gopark(0xc000088000?, 0xa40960?, 0x1?, 0x0?, 0x0?)
	/usr/local/go/src/runtime/proc.go:381 +0xd6 fp=0xc000061f70 sp=0xc000061f50 pc=0x527716
runtime.goparkunlock(...)
	/usr/local/go/src/runtime/proc.go:387
runtime.(*scavengerState).park(0xc32b40)
	/usr/local/go/src/runtime/mgcscavenge.go:400 +0x53 fp=0xc000061fa0 sp=0xc000061f70 pc=0x511bf3
runtime.bgscavenge(0x0?)
	/usr/local/go/src/runtime/mgcscavenge.go:628 +0x45 fp=0xc000061fc8 sp=0xc000061fa0 pc=0x5121c5
runtime.gcenable.func2()
	/usr/local/go/src/runtime/mgc.go:179 +0x26 fp=0xc000061fe0 sp=0xc000061fc8 pc=0x509126
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc000061fe8 sp=0xc000061fe0 pc=0x554941
created by runtime.gcenable
	/usr/local/go/src/runtime/mgc.go:179 +0xaa

goroutine 18 [finalizer wait]:
runtime.gopark(0x1a0?, 0xc32f80?, 0xe0?, 0x24?, 0xc000060770?)
	/usr/local/go/src/runtime/proc.go:381 +0xd6 fp=0xc000060628 sp=0xc000060608 pc=0x527716
runtime.runfinq()
	/usr/local/go/src/runtime/mfinal.go:193 +0x107 fp=0xc0000607e0 sp=0xc000060628 pc=0x5081c7
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc0000607e8 sp=0xc0000607e0 pc=0x554941
created by runtime.createfing
	/usr/local/go/src/runtime/mfinal.go:163 +0x45

rax    0x0
rbx    0x46d5
rcx    0x7f797b8b0844
rdx    0x6
rdi    0x46d2
rsi    0x46d5
rbp    0x7f7953401a70
rsp    0x7f7953401a30
r8     0xffffffff
r9     0x0
r10    0x8
r11    0x246
r12    0x7f79534026c0
r13    0x7f7953401b80
r14    0x6
r15    0x7f797b9c3531
rip    0x7f797b8b0844
rflags 0x246
cs     0x33
fs     0x0
gs     0x0
exit status 2

```

**Comment by gucio321** [comment]
I'm prettty sure that it happens due to something in font atlas...
I happens after beforeDestroy call while exiting cgo callbacks.
I have no idea what could it be and how to debug this...

**Comment by AllenDang** [comment]
> ```
> /home/me/git/giu/MasterWindow.go:222
> ```

What's the code in `/home/me/git/giu/MasterWindow.go:222`?

**Comment by gucio321** [comment]
@AllenDang the problem is that MasterWindow.go:222 is
```go
                Context.backend.Run(w.render)
```

**Comment by gucio321** [comment]
> It seems a texture has been freed twice.

I'ts a valuable tip. Maybe this is the problem?
https://github.com/AllenDang/cimgui-go/blob/606eb8908b6c3960a39b2c26cba733fc51014d45/texture.go#L37

**Comment by gucio321** [comment]
unfortunetly not :cry:

**Comment by gucio321** [comment]
> It seems a texture has been freed twice.

I think it isn't a problem. If I comment out `SetTexID call it still crashes after closing an app

**Comment by AllenDang** [comment]
> > It seems a texture has been freed twice.
> 
> I think it isn't a problem. If I comment out `SetTexID call it still crashes after closing an app

How about comment out CreateTexture?

**Comment by gucio321** [comment]
still the same

**Comment by Chillance** [comment]
> ugh and there is even a dedicated issue:
> 
> * [Panic when exiting with window manager help (e.g. ALT+F4) without SetBeforeDestroyContextHook after loading own fonts cimgui-go#90](https://github.com/AllenDang/cimgui-go/issues/90)

So, it's really a different issue not something that needs to be fixed before this? Would be a shame if that is a showstopper when so much work and progress has been made here and this can't be over the finish line because of that one.

**Comment by gucio321** [comment]
yo're right, this issue in cimgui-go is the most important one for me but unfortunetly I have no idea how to fix it

**Comment by gucio321** [comment]
Good news: I found the code causing the crash:
It is `giu.Set(Default)FontFromBytes` if using `Set(Default)Font` it deosn't crash.
I suppose this is because the second method passes only font file's path when the first one operates byte slice

**Comment by gucio321** [comment]
modified glfw_backend.go
```c
void igRunLoop(GLFWwindow *window, VoidCallback loop, VoidCallback beforeRender, VoidCallback afterRender,
               VoidCallback beforeDestroyContext) {
  glfwMakeContextCurrent(window);
  ImGuiIO *io = igGetIO();

  // Load Fonts
  // - If no fonts are loaded, dear imgui will use the default font. You can
  // also load multiple fonts and use igPushFont()/PopFont() to select them.
  // - AddFontFromFileTTF() will return the ImFont* so you can store it if you
  // need to select the font among multiple.
  // - If the file cannot be loaded, the function will return NULL. Please
  // handle those errors in your application (e.g. use an assertion, or display
  // an error and quit).
  // - The fonts will be rasterized at a given size (w/ oversampling) and stored
  // into a texture when calling ImFontAtlas::Build()/GetTexDataAsXXXX(), which
  // ImGui_ImplXXXX_NewFrame below will call.
  // - Read 'docs/FONTS.md' for more instructions and details.
  // - Remember that in C/C++ if you want to include a backslash \ in a string
  // literal you need to write a double backslash \\ !
  // io.Fonts->AddFontDefault();
  // io.Fonts->AddFontFromFileTTF("../../misc/fonts/Roboto-Medium.ttf", 16.0f);
  // io.Fonts->AddFontFromFileTTF("../../misc/fonts/Cousine-Regular.ttf", 15.0f);
  // io.Fonts->AddFontFromFileTTF("../../misc/fonts/DroidSans.ttf", 16.0f);
  // io.Fonts->AddFontFromFileTTF("../../misc/fonts/ProggyTiny.ttf", 10.0f);
  // ImFont* font =
  // io.Fonts->AddFontFromFileTTF("c:\\Windows\\Fonts\\ArialUni.ttf", 18.0f,
  // NULL, io.Fonts->GetGlyphRangesJapanese()); IM_ASSERT(font != NULL);

  printf("1\n");
  // Main loop
  double lasttime = glfwGetTime();
  while (!glfwWindowShouldClose(window)) {
  printf("2\n");
    if (beforeRender != NULL) {
      beforeRender();
    }

    glfw_render(window, loop);

    while (glfwGetTime() < lasttime + 1.0 / glfw_target_fps) {
      // do nothing here
    }
    lasttime += 1.0 / glfw_target_fps;

    if (extra_frame_count > 0) {
      extra_frame_count--;
    } else {
      glfwWaitEvents();
      extra_frame_count = MAX_EXTRA_FRAME_COUNT;
    }

    glfwPollEvents();

    if (afterRender != NULL) {
      afterRender();
    }
  printf("3\n");
  }

  // Cleanup
  printf("4\n");
  ImGui_ImplOpenGL3_Shutdown();
  printf("5\n");
  ImGui_ImplGlfw_Shutdown();

  printf("6\n");
  if (beforeDestroyContext != NULL) {
    beforeDestroyContext();
  }
  printf("7\n");

  igDestroyContext(0);
  printf("8\n");

  glfwDestroyWindow(window);
  printf("9\n");
  glfwTerminate();
  printf("10\n");

  return;
```
it prints up to 7 and then crashes

**Comment by gucio321** [comment]
```c
// This function is merely here to free heap allocations.
void ImGui::Shutdown()
{
    // The fonts atlas can be used prior to calling NewFrame(), so we clear it even if g.Initialized is FALSE (which would happen if we never called NewFrame)
    ImGuiContext& g = *GImGui;
    if (g.IO.Fonts && g.FontAtlasOwnedByContext)
    {
        g.IO.Fonts->Locked = false;
        IM_DELETE(g.IO.Fonts); // <- this line crashes
    }
```

**Comment by gucio321** [comment]
further stack trace:
IM_DELETE -> ~ImFontAtlas -> ClearInputData
```c
    for (int i = 0; i < ConfigData.Size; i++)
        if (ConfigData[i].FontData && ConfigData[i].FontDataOwnedByAtlas)
        {
            fprintf(stdout,"clear input data 1.1\n");
            IM_FREE(ConfigData[i].FontData); // <- this one
            fprintf(stdout,"clear input data 1.2\n");
            ConfigData[i].FontData = NULL;
            fprintf(stdout,"clear input data 1.3\n");
        }
```

**Comment by gucio321** [comment]
so I suppose we have to set `SetFontDataOwnedByAtlas(false)` but if I do so it crashes because "Font not loaded" (idk why?)

**Comment by Chillance** [comment]
Gotta love the progress here. Hopefully will this be resolved soon. In the meantime, I wanted to ask how can one make it easy to use the docking branch of imgui instead of the default one? Any thoughts put into how to easily change to use that?

**Comment by gucio321** [comment]
cimgui-go uses docking branch of imgui by default

**Comment by Chillance** [comment]
Oh, awesome! I guess there would be no big reason to have support to change and use master one then?

**Comment by gucio321** [comment]
_It isn't tested yet but_ it shouldn't be a big problem for cimgui-go to re-generate for imgui master.
I'm not sure about giu but if I'm right you'd just need to create a dedicated branch and change cimgui upstream.

**Comment by gucio321** [comment]
waiting for
- https://github.com/AllenDang/cimgui-go/pull/187
because it _might_ be related

**Comment by gucio321** [comment]
development update:
there are two news: good and bad one:
- one blocker is going to be removed :tada: https://github.com/AllenDang/cimgui-go/issues/169
- but unfortunetly there is another much worse one :cry: https://github.com/AllenDang/cimgui-go/issues/90 (ref https://github.com/golang/go/issues/62356)

**Comment by Chillance** [comment]
Why are these new issues showing up now? Is that because of updates to imgui which this needs to catch up to?

**Comment by gucio321** [comment]
Not exactly. It's partially because of some improvements in cimgui-go and partially because don't functions are missing yet

**Comment by gucio321** [comment]
### development status
- still having some issues with missing GLFW backend stuff (https://github.com/AllenDang/cimgui-go/issues/195)
- font atlas still broken
   - if FontDataOwnedByAtlas set to `true` (default) I'm getting panic message after closing app caused by conflict of GO's GC (Garbage COllector) and internal imgui's state (that tries to free font data pointer). In theory we should be able to do something with that using a new workaround addend in cimgui-go (https://github.com/AllenDang/cimgui-go/pull/211) But I'd prefer to set FontDataOwnedByAtlas to false.
   - If FontDataOwnedByAtlas set to false everything crashes immediately after opening app. I've opened a discussion in Dear ImGui's repo as I suppose it might be a bug in upstream: https://github.com/ocornut/imgui/discussions/6822

**Comment by gucio321** [comment]
## status update

- I managed to fix fonts issue :tada: 
- MacOS job works now

**Comment by gucio321** [comment]
### UPDATE

Now I'm working on implementation of Markdown widget (see [cimmarkdown](https://github.com/gucio321/cimmarkdown))

**Comment by Chillance** [comment]
I had no idea this was such a huge effort. And it seems to keep add things... Is there any way some things could be fixes after merging this?

**Comment by gucio321** [comment]
Yes. I can merge the current state to master (it mostly works) and keep adding missing things
- code editor
- markdown editor
- some plots

**Comment by AllenDang** [comment]
> Yes. I can merge the current state to master (it mostly works) and keep adding missing things
> 
> * code editor
> * markdown editor
> * some plots

Thanks so much for your effort, I think we should do the migration now, with a new release, and add more fixes later.

**Comment by gucio321** [comment]
@AllenDang I'm going to perform the following actions:
- create release from current master
- create a clone of matter and call it e.g. imgui-go
- merge this to master

**Comment by Chillance** [comment]
Wow! This is big deal! Thank you all for your efforts!

**Comment by gucio321** [comment]
finally after.... **9.5 months** :tada:

**Pull request merged**